napari/plugins/_npe2.py

```diff
@@ -488,29 +504,183 @@ def _add_sample(
     return submenu, sample_actions
 
 
+def _get_widget_param(widget_callable, widget_name) -> str:
+    """Get widget parameter name (if any) and check type."""
+    from napari.viewer import Viewer
+
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer':
+                    widget_param = param.name
+                    break
+                if param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    elif isinstance(widget_callable, MagicFactory) or callable(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "Widgets must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or callable, but {widget} is of type {type_}. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+                type_=type(widget_callable),
+            )
+        )
+    return widget_param
+
+
+def _get_widgets_submenu_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Any], List[Any]]:
+    """Get widget submenu and actions for a single npe2 plugin manifest."""
+    from napari._app_model.constants import MenuGroup, MenuId
+    from napari.plugins import _npe2, menu_item_template
+    from napari.viewer import Viewer
+
+    if TYPE_CHECKING:
+        from napari._qt.qt_main_window import Window
+
+    # If no widgets, return
+    if not mf.contributions.widgets:
+        return [], []
+
+    widgets = mf.contributions.widgets
+    multiprovider = len(widgets) > 1
+    submenu_id, submenu = _get_multiprovider_submenu(
+        multiprovider,
+        MenuId.MENUBAR_PLUGINS,
+        mf,
+        MenuGroup.PLUGIN_CONTRIBUTIONS,
+    )
+
+    widget_actions: List[Action] = []
+    for widget in widgets:
+        full_name = menu_item_template.format(
+            mf.display_name, widget.display_name
+        )
+
+        def _widget_callback(
```

Comment:
I think we should move the widget callback out into it's own function. If we need any locals we should pass them to an enclosing function that returns us the `_widget_callback`. It's too hard to get the flow of this function otherwise imo